### ISSUE TITLE
robot case only generator output.xml but not generator log.html and report.html

### ISSUE THREAD

**Issue by antzjm** [robotframework/robotframework#4040]
In my robotframe case, the robot command line only generate **output.xml** and do not generator log.html and report.html.
But I can not see any error info. So I raise this issue for robot framework team for help.

**Comment by pekkaklarck** [comment]
How is Robot called?

**Comment by antzjm** [comment]
> How is Robot executed? Is it possible log and report generation is explicitly disabled?

Here is my command line, `robot -L trace -s accept2 -v platform:NCS -v NAMESPACE:cran1 -e requires-Intel -e requires-NCIR -e requires-StarlingX /usr/libexec/ContainerInfraTest/`.

**Comment by pekkaklarck** [comment]
Ok, log and report generation isn't explicitly disabled, Do you y get any errors?

**Comment by pekkaklarck** [comment]
Any new information related to this? I asked about possible errors above, but now notice that in the original description you mentioned you don't get any. Unfortunately with this information alone it's not possible to debug this any further.

**Comment by gauravgandhi1315** [comment]
I am facing the same issue

**Comment by pekkaklarck** [comment]
With this information alone I cannot debug this any further.

**Comment by buckleyGI** [comment]
Hi @pekkaklarck We are facing the same issue: the output file is generated but not the log.html and report.html

It happens for us when we use **re**bot. The command in essence is

`rebot --outputdir=output --output output.xml --merge 1.xml 2.xml .. 44.xml`

(so as simple as can get)

This command executed fine when run with v3 but now fails v4.
I also think it has to do with the number of tests that, when high, makes the log+report fail.

This is the output we get
`Output:  /usr/src/project/output/output.xml`

Wile this is expected
```
Output:  /usr/src/project/output/output.xml
Log:     /usr/src/project/output/log.html
Report:  /usr/src/project/output/report.html
```

We execute rebot in azure devops and don't see any errors that would indicated a problem during log and report.
It just fails and continues silently it seems.

Eager to help out here.

Is there anything we can do to gather more information like capturing the rebot processing by a log or increasing the verbosity level?

**Comment by buckleyGI** [comment]
This issue also seem related to our case
https://github.com/robotframework/RIDE/issues/1871

**Comment by pekkaklarck** [comment]
Do you mean Robot Framework versions with v3 and v4? Do you get any error or is the only symptom that log and report aren't generated?

It would be great if someone experiencing this problem could create a simple example demonstrating the problem. If I could reproduce the issue locally, I could easily debug it further. And if there turns out to be a bug in RF, I could most likely easily fix that as well.

Notice that we are planning to create RF 4.1.2 release this Thursday and the plan is that it would be the last RF 4 release. If the is a bug in Robot, it would be great to get it fixed and fix included in that release. I can postpone the release a little if there's new information about this issue.

**Comment by buckleyGI** [comment]
> Do you mean Robot Framework versions with v3 and v4? Do you get any error or is the only symptom that log and report aren't generated?

Yes, we upgraded RF from v3 to v4. We do not get any error. Output is generated, log and report not.

> It would be great if someone experiencing this problem could create a simple example demonstrating the problem. If I could reproduce the issue locally, I could easily debug it further. And if there turns out to be a bug in RF, I could most likely easily fix that as well.

We have about 1500 tests that result in an output.xml of 700 Meg.
Since it seems to be a problem with the number of tests a simple repro is not possible.
Packaging up these huge files and send it over woth some medium is not possible short term.

We are learning more about the problem and I am currently trying to determine at what number of tests the files faild to generate.
If rebot is run in azure devops it fails to generate these fails but if I run the docker locally it does generate them.
This indicates that the docker host and the resources it give to the container triggers the problem.

But we are running totally blind as there is no error message whatsoever. Robot fails silently.
We don't have a clue what is going wrong.

I didn't know that the v4 branch will stop. Does that mean that V5 is in the works? Maybe there is a communication on this? Thank you.

**Comment by boakley** [comment]
I bet you could write a short python script that automatically generates
the necessary number of tests in a temporary directory, and then runs robot
against the generated suite of tests.

**Comment by buckleyGI** [comment]
Pinpointing the breaking point we see that the 2 files stop being generated when we cross about 1300 tests and the output.xml becomes about 700 Megabytes.

**Comment by buckleyGI** [comment]
Producing an isolated repro case seems practically impossible. Is there a way we can debug this issue on our side?

**Comment by buckleyGI** [comment]
We are currently trying to mitigate the issue by splitting the workload in 2 and merging these 2 intermediate results later.

Before
```
rebot --merge file01.xml        ... file43.xml        --ouput output.xml
```
After
```
rebot --merge file01.xml        ... file22.xml        --ouput intermidiate1.xml
rebot --merge file23.xml        ... file43.xml        --ouput intermidiate2.xml
rebot --merge intermidiate1.xml intermidiate2.xml     --ouput output.xml

```
I don't know this will make a difference but its the last solution we could come up with.

**Comment by buckleyGI** [comment]
Splitting the workload in 2 and merging the intermediate results does not work either. We are out of options.
Is there a way to let rebot log so that we can possible get at the reason why it doesn't produce the log and report files?

**Comment by pekkaklarck** [comment]
Rebot doesn't have any hard limit for tests or output file sizes. Because you are dealing with huge files, the likely root cause is Rebot running out of memory. It's strange there's no error, but strange things happen when memory runs out. Could you try increasing the amount of memory on the machine where this occurs to verify my assumption?

If I'm right about this being a memory issue, there's clearly a regression related to memory usage between RF 3 and 4. Which exact RF version still worked and which doesn't?

**Comment by pekkaklarck** [comment]
As I wrote earlier, the plan was to create RF 4.1.2 this week and it to be the last RF 4 release. The latter is still true, but we can postpone RF 4.1.2 to get this issue fixed as well. That will require still more help in debugging the problem. If this is confirmed to be a memory issue and I know which is still a good version, I can try creating huge output files and profile the memory usage myself.

**Comment by buckleyGI** [comment]
Before
```
robotframework==3.2.2
robotframework-seleniumlibrary==4.5.0
robotframework-databaselibrary==1.2
robotframework-requests==0.7.1
robotframework-pabot==1.10
```
After
```
robotframework==4.1
robotframework-seleniumlibrary==5.1.3
robotframework-databaselibrary==1.2.4
robotframework-requests==0.9.1
robotframework-pabot==2.0.1
```

**Comment by buckleyGI** [comment]
> Could you try increasing the amount of memory on the machine where this occurs to verify my assumption?

Going at the infra layer to make changes there (increase ram/cpu) was on my radar.
We are already creating subreports that when added together represent all tests. So the global report that we generated in the past is not strictly needed and we are going to disable it.

This will deblock our current process so that has first priority now. We still extracted some information out of report.html so it still requires a rewrite of logic.

After that I can do some more trials to get the volume problem fixed, happy to assist in any way, but other backlog work will get prio. I gave it a window of 3 days to get fixed and I suspect any further investigation will become more complex but I think I can still repro after deblocking.

**Comment by pekkaklarck** [comment]
Any possibility you could test with RF 4.0.2? With other tools it would be better to use same versions "after" than you used "before". I doubt that it has an effect but narrowing this down would still be good.

**Comment by buckleyGI** [comment]
> Any possibility you could test with RF 4.0.2?
Why this specific version?

**Comment by pekkaklarck** [comment]
Yeah, I'd like to know at what point the problem occurred in more detail. Makes it easier to debug it.

**Comment by pekkaklarck** [comment]
I created some example tests and compared speed and memory usage between 3.2, 4.0 and 4.1. I noticed memory usage got roughly 10% up in 4.0 and speed got a bit slower as well. I profiled performance using cProfile and was able to enhance it slightly. Those changes didn't seem to have affect on the memory usage, though.

I don't have experience using such memory profilers that would give detail where exactly the memory is spent so that I could compare results between different versions. `/usr/bin/time -v` gives maximum memory usage, though, and using that and git bisect I was able to find that overall memory usage raised in commit 21c6b468a6f19181e44300c2be5a3fa4afd28471. Need to next investigate that commit in more detail and see could code be enhanced in this regard.

**Comment by buckleyGI** [comment]
Thanks for investigating this so methodologically. Bisecting sounds a bit labor intensive but is the way to go about with these performance problems indeed.

I am currently rewriting our process so that it supports both scenarios: 
- either the 2 files get generated
- or they don't and we have a fall back mechanism that will continue with the subreprots

And what do you know since yesterday the process generates the 2 files without manual intervention or noticeable change in input (!). So this reaffirms the superspion that its resource related.

Still, we don't get an error message when it failed so we cannot ignore this (temp fixed) problem. And I am sure this problem will resurface. 

If you or I come up with an idea to trigger the problem again then lets do that so we can make rebot fail hard or, if possible, let it say by an error message what went wrong and how to remedy it (eg. I had x RAM but needed y RAM).
I am more familiar with .NET and I know that an out of memory exception can be nasty for error handling but I believe it handles it well. .NET has a problem with a stackoverflow, here I observed it would also fail without proper error handling.
So not only OOM but stackoverflow is a candidate of the problem. Still, why it now suddenly starts working I have no clue.

**Comment by pekkaklarck** [comment]
Did it succeed just once or are all runs now successful?

There not being any error is very annoying. I think there should be something even if memory runs out, but it could be something pretty generic and it's possible we'd hide the traceback. It would be easy to change the code so that traceback is shown, but since there's no error in the first place I seriously doubt it would matter. Is there any change you have configured stderr to be redirect somewhere? That would be pretty logical explanation for not seeing any errors.

**Comment by pekkaklarck** [comment]
Investigation I did unfortunately didn't really reveal anything that could be easily changed to improve memory usage. The commit 21c6b468a6f19181e44300c2be5a3fa4afd28471 that I mentioned above changes `TestCase` objects so that they have `setup` and `teardown` directly and that they are concrete `Keyword` objects. This thus requires creating two new `Keyword` objects per test. I tested that there's a pretty clear memory increase when creating 10000 `TestCase` objects directly in Python, but the overall memory usage is still so low it's hard to believe this would cause any real problems. The benefits of these model changes, explained in #3749, are also pretty big and changing the model again would be backwards incompatible.

While investigating this, I noticed few places where I could do micro enhancements to performance (see 01510f25a190b56329c17146fff2cc52240442ad for details). Nothing big enough to warrant a separate issue or a mention in RF 4.1.2 release notes but nice anyway. Most likely these changes won't affect memory usage at all.

Because investigation didn't reveal anything that could be easily fixed in RF 4.1.2, I'll no create RF 4.1.2 rc 1 and possible memory usage improvements need to wait for future releases. Let me know @antzjm if you find any more information about this!
